Fix relative links to api_reference having trailing slash and 404

frontsideair opened the pull request:
### Related Issues/PRs

<!-- Uncomment 'Resolve' if this PR can close the linked items. -->

### What changes are proposed in this pull request?

Use `<APILink>` component everywhere a link is needed to `api_reference` and use plain `<a>` tag to avoid trailing slash insertion.

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [x] `area/docs`: MLflow documentation pages
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/prompt`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

<!-- patch -->

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

frontsideair committed:
Use APILink component for links to api_reference to avoid trailing slash

Signed-off-by: Fatih Altinok <fatihaltinok@live.com>

frontsideair committed:
Add empty hash to api link

Signed-off-by: Fatih Altinok <fatihaltinok@live.com>

harupy commented on docs/src/components/APILink/index.tsx:
```diff
@@ -16,17 +16,17 @@ const getModule = (fn: string) => {
   }
 
-  const docLink = useBaseUrl(`/${APIModules[module]}#${fn}`);
+  const docLink = `https://mlflow.org/docs/latest/${APIModules[module]}#${hash ?? fn}`; // FIXME
```
Does this work in the preview?

frontsideair committed:
Revert doclink change made for dev only

Signed-off-by: Fatih Altinok <fatihaltinok@live.com>

TomeHirata commented on docs/src/components/APILink/index.tsx:
```diff
@@ -25,8 +25,8 @@ export function APILink({ fn, children }: { fn: string; children?: React.ReactNo
   const docLink = useBaseUrl(`/${APIModules[module]}#${fn}`);
   return (
-    <Link to={docLink} target="_blank">
+    <a href={docLink} target="_blank">
```
I'm not sure if using a tag is the right approach here. Won't we lose the ability to detect broken links when we use a tag? Instead, can't we redirect api doc links with trailing slash or consolidate the path policy (display api doc correctly with trailing slash)?

frontsideair replied:
I believe broken links check only works for Docusaurus internal links, not API reference links. I can add a redirect too, to catch any missing links, but I think it would be more consistent to use `<APILink>` component for all links to API reference.

Also I should probably add a lint rule to use the component for API reference links.

frontsideair replied:
I reverted back to using `<Link>` and intentionally broke a link to check, didn't get a broken links error during build. Seems like we're not losing any safety here.

TomeHirata replied:
Got it. Let's go with `a` tag then. Thanks for checking

frontsideair committed:
Add eslint rule to prefer APILink component

Signed-off-by: Fatih Altinok <fatihaltinok@live.com>

TomeHirata reviewed (approved):
LGTM

TomeHirata commented on docs/src/api_modules.json:
```diff
@@ -53,5 +53,11 @@
-  "mlflow.server.auth": "api_reference/auth/python-api.html"
-}
\ No newline at end of file
+  "mlflow.server.auth": "api_reference/auth/python-api.html",
```
Btw, should this file be tracked in git? I saw that this file is in .gitignore cc: @daniellok-db 

https://github.com/mlflow/mlflow/blob/de182788be8721d557b90784f1064d255c6306a1/.gitignore#L41

frontsideair replied:
I added the manual overrides to the `update-api-modules.ts` so they won't get overwritten.

daniellok-db replied:
Looks good!

harupy commented on docs/src/components/APILink/index.tsx:
```diff
@@ -25,8 +25,8 @@ export function APILink({ fn, children }: { fn: string; children?: React.ReactNo
 
   const docLink = useBaseUrl(`/${APIModules[module]}#${fn}`);
```
The API link has two `latest`:

<img width="584" alt="image" src="https://github.com/user-attachments/assets/63fe54a2-0a4a-4977-ba23-53cadc39a230" />

frontsideair replied:
I also noticed this earlier this week for the preview docs, I'll land a separate PR for it.

frontsideair replied:
@harupy https://github.com/mlflow/mlflow/pull/16676

daniellok-db commented on docs/eslint-plugin-mlflow-docs/rules/prefer-apilink-component.js:
```diff
@@ -0,0 +1,89 @@
+ *
+ * @type {import('eslint').Rule.RuleModule}
+ */
+module.exports = {
```
this is great!! thank you

daniellok-db commented on docs/src/components/APILink/index.tsx:
```diff
@@ -16,7 +16,7 @@ const getModule = (fn: string) => {
 };
 
-export function APILink({ fn, children }: { fn: string; children?: React.ReactNode }) {
+export function APILink({ fn, children, hash }: { fn: string; children?: React.ReactNode; hash?: string }) {
```
ah i see, there are section headers from non-functions in the API docs as well. that makes sense. however it seems unused here?

frontsideair replied:
I think I messed something up when reverting a dev only thing, thanks for catching!

frontsideair committed:
Fix bad revert, use hash when it exists in APILink

Signed-off-by: Fatih Altinok <fatihaltinok@live.com>

frontsideair committed:
Add new docs links to generator function as manual overrides

Since `yarn update-api-modules` script overwrites `api_modules.json`,
we should include them in the generator so they won't get overwritten

Signed-off-by: Fatih Altinok <fatihaltinok@live.com>

daniellok-db reviewed (approved):
LGTM!
